metloom/pointdata/files.py

```diff
@@ -0,0 +1,253 @@
+                    with open(filename, mode='w+') as fp:
+                        for line in r.iter_lines():
+                            fp.write(line.decode('utf-8') + '\n')
+        return filenames
+
+    def _get_one_variable(self, resp_df, period, variable: SensorDescription):
+        """
+        Retrieve a single variable and process it accordingly
+        """
+        method = "sum" if variable.accumulated else "average"
+        if self._verify_sensor(resp_df, variable):
+            isolated = resp_df.loc[:, variable.code]
+
+            # TODO: This may only be true for SNOWEX
+            isolated.loc[isolated == -9999] = np.nan
+            if method == 'average':
+                data = isolated.resample(period).mean()
+            elif method == 'sum':
+                data = isolated.resample(period).sum()
+            else:
+                raise Exception('Invalid aggregation method')
+        else:
+            data = None
+            LOG.debug(f"No data returned for {variable}")
+
+        return data
+
+    def _get_data(self, start_date, end_date, variables: List[SensorDescription],
+                  period):
+        """
+        Utilizes cached data or downloads the data
+        """
+        self.valid = self._verify_station()
+
+        if not self.valid:
+            return None
+
+        urls = self._file_urls(self._station_info.station_id,
+                               start_date, end_date)
+
+        # Make the cache dir
+        if not self._cache.is_dir():
+            os.mkdir(self._cache)
+
+        # Download data if it doesn't exist locally.
+        files = self._download(urls)
+        dfs = [pd.read_csv(f, index_col=False, low_memory=False) for f in files]
+        resp_df = pd.concat(dfs)
+        resp_df = self._assign_datetime(resp_df)
+
+        # use a predifined index to show nans in the event of patch data
+        df = pd.DataFrame(index=pd.date_range(start_date, end_date, freq=period,
+                                              name='datetime'),
+                          columns=[v.name for v in variables])
+
+        # Use this instead .loc to avoid index on patchy data
+        ind = (resp_df.index >= start_date) & (resp_df.index < end_date)
+        isolated = resp_df.loc[ind, resp_df.columns]
+        for i, variable in enumerate(variables):
+            df_var = self._get_one_variable(isolated, period, variable)
+            if df_var is not None:
+                if not np.all(df_var.isnull()):
+                    df.loc[df_var.index, variable.name] = df_var
+
+        # All nan data suggests no matching data
+        if np.all(df.isnull()):
+            return None
+
+        # Set the site info
+        df["site"] = [self.id] * len(df)
+        df["datasource"] = [self.DATASOURCE] * len(df)
+        # Make this a geodataframe
+        df = gpd.GeoDataFrame(df, geometry=[self.metadata] * len(df))
+        df = df.reset_index().set_index(["datetime", "site"])
+
```

Comment:
Can you call `self.validate_sensor_df(df)` before returning? This checks some of the column info so that we keep consistent formatting.

Reply:
You got it.